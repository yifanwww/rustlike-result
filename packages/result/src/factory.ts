import type { Result } from './Result';
// eslint-disable-next-line import/no-cycle
import { RustlikeResult } from './RustlikeResult';

/**
 * Creates a `Result` that contains the success value.
 *
 * Examples:
 * ```ts
 * const result1 = Ok(1);
 * const result2 = Ok<number, string>(1);
 * const result3: Result<number, string> = Ok(1);
 * ```
 */
export function Ok<T, E = never>(value: T): Result<T, E> {
    return RustlikeResult.Ok(value);
}

/**
 * Creates a `Result` that contains the error value.
 *
 * Examples:
 * ```ts
 * const result = Err('Some error message');
 * ```
 */
export function Err<E>(error: E): Result<never, E>;
/**
 * Creates a `Result` that contains the error value.
 *
 * Examples:
 * ```ts
 * const result1 = Err<number, string>('Some error message');
 * const result2: Result<number, string> = Err('Some error message');
 * ```
 */
export function Err<T, E>(error: E): Result<T, E>;
export function Err<T, E>(error: E): Result<T, E> {
    return RustlikeResult.Err(error);
}
