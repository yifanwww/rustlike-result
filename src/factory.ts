import { RustlikeResult } from './result';
import type { Result } from './types';

/**
 * Creates a `Result` that contains the success value.
 *
 * Examples:
 * ```ts
 * const result = Ok(1);
 * ```
 */
export function Ok<T>(value: T): Result<T, never>;
/**
 * Creates a `Result` that contains the success value.
 *
 * Examples:
 * ```ts
 * const result1 = Ok<number, string>(1);
 * const result2: Result<number, string> = Ok(2);
 * ```
 */
export function Ok<T, E>(value: T): Result<T, E>;
export function Ok<T, E>(value: T): Result<T, E> {
    return RustlikeResult.Ok(value);
}

/**
 * Creates a `Result` that contains the error value.
 *
 * Examples:
 * ```ts
 * const result = Err('Some error message');
 * ```
 */
export function Err<E>(error: E): Result<never, E>;
/**
 * Creates a `Result` that contains the error value.
 *
 * Examples:
 * ```ts
 * const result1 = Err<number, string>('Some error message');
 * const result2: Result<number, string> = Err('Some error message');
 * ```
 */
export function Err<T, E>(error: E): Result<T, E>;
export function Err<T, E>(error: E): Result<T, E> {
    return RustlikeResult.Err(error);
}
