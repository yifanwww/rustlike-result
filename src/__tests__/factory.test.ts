import { Err, Ok } from '../factory';
import type { Result } from '../types';

describe(`Test fn \`${Ok.name}\``, () => {
    it('should create `Ok` result', () => {
        const result1 = Ok(1);
        const result2 = Ok<number, string>(1);
        const result3: Result<number, string> = Ok(2);

        expect(result1).toMatchSnapshot();
        expect(result2).toMatchSnapshot();
        expect(result3).toMatchSnapshot();
    });
});

describe(`Test fn \`${Err.name}\``, () => {
    it('should create `Err` result', () => {
        const result1 = Err('Some error message');
        const result2 = Err<number, string>('Some error message');
        const result3: Result<number, string> = Err('Some error message');

        expect(result1).toMatchSnapshot();
        expect(result2).toMatchSnapshot();
        expect(result3).toMatchSnapshot();
    });
});
