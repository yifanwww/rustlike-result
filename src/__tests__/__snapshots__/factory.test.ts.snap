// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test fn \`Err\` should create \`Err\` result 1`] = `
RustlikeResult {
  "_error": "Some error message",
  "_type": "err",
  "_value": undefined,
}
`;

exports[`Test fn \`Err\` should create \`Err\` result 2`] = `
RustlikeResult {
  "_error": "Some error message",
  "_type": "err",
  "_value": undefined,
}
`;

exports[`Test fn \`Err\` should create \`Err\` result 3`] = `
RustlikeResult {
  "_error": "Some error message",
  "_type": "err",
  "_value": undefined,
}
`;

exports[`Test fn \`Ok\` should create \`Ok\` result 1`] = `
RustlikeResult {
  "_error": undefined,
  "_type": "ok",
  "_value": 1,
}
`;

exports[`Test fn \`Ok\` should create \`Ok\` result 2`] = `
RustlikeResult {
  "_error": undefined,
  "_type": "ok",
  "_value": 1,
}
`;

exports[`Test fn \`Ok\` should create \`Ok\` result 3`] = `
RustlikeResult {
  "_error": undefined,
  "_type": "ok",
  "_value": 2,
}
`;
